docs/design/object-locking.md

```diff
@@ -0,0 +1,187 @@
+be set
+
+### Retention Modes
+
+Both Governance and Compliance modes retain lock on an object until the set
+retention period.
+
+#### Governance mode
+
+In addition to preventing deletions on an object version by users,
+`GOVERNANCE` mode allows delegating a permission to certain users so that they
+can override the lock settings, for example changing the retention mode, period
+or deleting the object altogether.
+Either the root (account) or a user with `s3:BypassGovernanceRetention`
+permission can send a delete request with
+`x-amz-bypass-governance-retention:true` header to override and delete the
+object.
+
+#### Compliance mode
+
+When a lock is placed on an object using `COMPLIANCE` mode, the object version
+cannot be deleted by any user until the retention period expires.
+This includes root user (account credentials) in the account and no other user
+can be given permission either to override the settings or delete the version.
+
+### Retention Period
+
+Retention period defines the term until when the object is protected from
+deletes.
+Retention period can be set on bucket level, which acts as a default for all
+objects put in the bucket after the setting is applied. The default retention
+period set on a bucket can be overriden on the object level by setting the date
+and time using the header `x-amz-object-lock-retain-until-date`.
+
+**Buckets** -  retention period can be set in either `days` or `years`, but not
+both at the same time.
+
+**Objects** - retention period can be set as date and time when the object is
+expected to expire.
+
+### Legal Hold
+
+Legal hold can be enabled on an object version. Once a legal hold is enabled,
+regardless of the object's retention date or retention mode, the object version
+cannot be deleted until the legal hold is removed.
+Legal hold can be set on an object version during PUT Object request by setting
+`x-amz-object-lock-legal-hold` header or using PUT Object Legal Hold api
+request.
+Root users with account credentials or IAM users who are given the permission
+`s3:PutObjectLegalHold` are allowed to set Legal hold on an object version.
+
+### Implementation
+
+#### Storing Object lock configuration
+
+Object lock can be enabled on a bucket during bucket creation using
+`x-amz-bucket-object-lock-enabled` header. This is stored along with the
+bucket's metadata defined in BucketInfo model.
+Any default lock settings (governance/compliance mode, days/years for retention
+period) defined using `PutObjectLockConfiguration` request are stored along
+with the Bucket's metadata.
+When a PUT Object request is received, the lock configuration is evaluated and
+retain until date and time is calculated and set as `retain-until-date`
+property on the object's metadata.
+Lock configuration on a PUT object request is evaluated the following order
+
+1. Bucket's configuration for retention mode/period
+2. Object's configuration in the PUT Object for retention mode/period
```

Comment:
PUT Object Retention?

Reply:
I meant lock configuration that can be set on a PUT object request for example with header `x-amz-object-lock-retain-until-date`